Add docker protocol support for llama-server model loading

ericcurtin opened the pull request:
To pull and run models via: llama-server -dr gemma3
Add some validators and sanitizers for Docker Model urls and metadata

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +742,192 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+static bool common_docker_download_blob(const std::string & blob_url,
+                                        const std::string & token,
+                                        const std::string & local_path) {
+    curl_ptr       curl(curl_easy_init(), &curl_easy_cleanup);
+    curl_slist_ptr http_headers;
```
Can you elaborate on cringe?

ericcurtin replied:
Not sure what you mean, this line defines the headers variable which is required for the pull to work, tested this, works great.

I did rewrite the code to make this variable declaration closer to where it's actually used.

CISC replied:
I'll illustrate with an example.

This is the intended readability formatting:
```cpp
type_t variable = "value";
tp_t   var      = "val";
```

But to match the readability with initializers you'd have to format it like this:
```cpp
type_t variable ("value");
tp_t   var      ("val");
```

Otherwise the variable gets drowned out in varying syntax.

ericcurtin replied:
I moved the variable so the alignment issue should be gone.

If we have issues with the alignment we should change the clang-format file, this wastes a lot of time in code reviews, every reviewer has different styles or nitpicks that they think is most correct and one cannot read another reviewers mind. Style is subjective.

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +742,192 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+    std::string model_spec = docker_url.substr(9);  // Remove "docker://"
+
+    // Parse ai/smollm2:135M-Q4_K_M
+    size_t      colon_pos = model_spec.find(':');
+    std::string repo, tag;
```
How the hell was I to interpret "I mean" as this? I would appreciate more clear feedback

ericcurtin commented on common/arg.cpp:
```diff
@@ -244,6 +244,10 @@ static bool curl_perform_with_retry(const std::string & url, CURL * curl, int ma
     return false;
 }
 
+struct FILE_deleter {
+    void operator()(FILE * f) const { fclose(f); }
+};
```
This code leads to less duplication as FILE_deleter is used in two functions

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +742,192 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+    common_remote_params params;
+    auto res = common_remote_get_content(url, params);
+
+    if (res.first != 200) {
+        throw std::runtime_error("Failed to get Docker registry token, HTTP code: " + std::to_string(res.first));
```
Changed

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +742,192 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+        manifest_params.headers.push_back(
+            "Accept: application/vnd.docker.distribution.manifest.v2+json,application/vnd.oci.image.manifest.v1+json");
+        auto manifest_res = common_remote_get_content(manifest_url, manifest_params);
+        if (manifest_res.first != 200) {
+            throw std::runtime_error("Failed to get Docker manifest, HTTP code: " + std::to_string(manifest_res.first));
```
Changed

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +742,192 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+        }
+
+        std::string manifest_str(manifest_res.second.begin(), manifest_res.second.end());
+        json        manifest = json::parse(manifest_str);
```
Changed

CISC commented on common/arg.cpp:
```diff
@@ -745,6 +742,192 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+            for (const auto & layer : manifest["layers"]) {
+                if (layer.contains("mediaType")) {
+                    std::string media_type = layer["mediaType"].get<std::string>();
+                    if (media_type == "application/vnd.docker.ai.gguf.v3" ||
+                        media_type.find("gguf") != std::string::npos) {
```
It's redundant though, the `find` matches the first. Matching just on `gguf` may be too lax, but I don't know what media types there may be for GGUF.

ericcurtin replied:
Changed just to check media_type

ericcurtin commented:
@JohannesGaessler @slaren PTAL

ericcurtin commented:
@ggerganov @JohannesGaessler @slaren @danbev struggle to get this reviewed, if you guys have cycles I'd appreciate it.

ericcurtin commented on common/common.cpp:
```diff
@@ -899,10 +900,13 @@ struct common_init_result common_init_from_params(common_params & params) {
     auto mparams = common_model_params_to_llama(params);
 
-    llama_model * model = llama_model_load_from_file(params.model.path.c_str(), mparams);
+    // Resolve Docker URLs if needed
+    std::string resolved_model_path = common_docker_resolve_model(params.model.path);
+
```
Will see what I can do!

ericcurtin commented:
I'm also having a change of heart, thinking of changing to a:

-d/--docker

at least the string would be the same then as suggested in dockerhub and the one used in docker model runner. It would also be more consistent with the huggingface approach.

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +742,206 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+static bool common_docker_download_blob(const std::string & blob_url,
+                                        const std::string & token,
+                                        const std::string & local_path) {
+    curl_ptr curl(curl_easy_init(), &curl_easy_cleanup);
```
Done

ericcurtin committed:
Add docker protocol support for llama-server model loading

To pull and run models via: llama-server -dr gemma3
Add some validators and sanitizers for Docker Model urls and metadata

Signed-off-by: Eric Curtin <eric.curtin@docker.com>

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +765,97 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+        return local_path;
+    } catch (const std::exception & e) {
+        LOG_ERR("Docker Model download failed: %s\n", e.what());
+        throw;
```
Done.

ericcurtin commented on common/arg.cpp:
```diff
@@ -745,6 +765,97 @@ std::pair<long, std::vector<char>> common_remote_get_content(const std::string &
+        std::replace(model_filename.begin(), model_filename.end(), '/', '_');
+        model_filename += "_" + tag + ".gguf";
+        std::string local_path = fs_get_cache_file(model_filename);
+
+        // Download the blob using common_download_file_single with is_docker=true
+        std::string blob_url = "https://registry-1.docker.io/v2/" + repo + "/blobs/" + gguf_digest;
```
Added some validators and sanitizers

ericcurtin replied:
validate_oci_digest is added, what other kind of concerns would you like us to address @ggerganov ?

ggerganov reviewed (commented):
The docker-related functions seem ok.

I'm not confident about the changes to `common_download_file_single` to support resumable downloads. Either wait for someone to review this part in details, or move it to a separate PR.
